File: src/scripts/tokenBucket.lua
-- tokenBucket.lua
-- KEYS[1]: key (e.g., rl:ip:127.0.0.1)
-- ARGV[1]: capacity (max tokens)
-- ARGV[2]: refill rate (tokens per second)
-- ARGV[3]: current timestamp (seconds)
-- ARGV[4]: requested tokens

local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local refill_rate = tonumber(ARGV[2])
local now = tonumber(ARGV[3])
local requested = tonumber(ARGV[4])

-- Get current tokens and last refill time
local info = redis.call('HMGET', key, 'tokens', 'last_refill')
local current_tokens = tonumber(info[1])
local last_refill = tonumber(info[2])

if current_tokens == nil then
  current_tokens = capacity
  last_refill = now
end

-- Refill tokens
local elapsed = now - last_refill
if elapsed > 0 then
  local new_tokens = elapsed * refill_rate
  current_tokens = math.min(capacity, current_tokens + new_tokens)
  last_refill = now
end

-- Check if enough tokens
local allowed = 0
if current_tokens >= requested then
  current_tokens = current_tokens - requested
  allowed = 1
end

-- Update Redis
redis.call('HMSET', key, 'tokens', current_tokens, 'last_refill', last_refill)
-- Set expiry to avoid stale keys (e.g., 2 * capacity / refill_rate or a safe default)
redis.call('EXPIRE', key, math.ceil(capacity / refill_rate * 2)) -- Prevent from one time visitors

-- Return: allowed, remaining tokens, reset time (not strictly applicable, but returning 0 for consistency or calculated time to full)
return {allowed, current_tokens}

File: src/strategies/TokenBucket.js
const fs = require('fs');
const path = require('path');
const BaseStrategy = require('./BaseStrategy');

const LUA_SCRIPT = fs.readFileSync(path.join(__dirname, '../scripts/tokenBucket.lua'), 'utf8');

class TokenBucket extends BaseStrategy {
    constructor(redisClient, keyPrefix) {
        super(redisClient, keyPrefix);

        // Define the custom command
        if (!this.redis.status || this.redis.status === 'ready') {
            this.defineCommand();
        } else {
            this.redis.on('ready', () => this.defineCommand());
        }
    }

    defineCommand() {
        // Check if command already exists to avoid errors on potential re-instantiation
        if (!this.redis.tokenBucket) {
            this.redis.defineCommand('tokenBucket', {
                numberOfKeys: 1,
                lua: LUA_SCRIPT,
            });
        }
    }

    /**
     * @param {string} identifier
     * @param {object} options { limit: number, window: number, cost: number }
     * limit: max tokens (capacity)
     * window: refill period (seconds) - implied rate = limit / window ? No, effectively rate
     * Actually, let's interpret options as:
     * capacity: max burst
     * refillRate: tokens added per second
     * OR standard: limit requests per window seconds.
     * Let's stick to standard: limit (capacity) requests per window (seconds).
     * Refill rate = limit / window.
     */
    async hit(identifier, options) {
        const { limit, window } = options;
        const cost = options.cost || 1;

        const key = `${this.keyPrefix}tb:${identifier}`;
        const refillRate = limit / window;
        const now = Date.now() / 1000; // seconds

        // Call lua script
        // ARGV[1]: capacity
        // ARGV[2]: refill rate
        // ARGV[3]: current timestamp
        // ARGV[4]: requested tokens (cost)

        // Note: defineCommand adds the method to the redis instance.
        // Ensure it's defined.
        if (!this.redis.tokenBucket) {
            this.defineCommand();
        }

        const result = await this.redis.tokenBucket(
            key,
            limit,
            refillRate,
            now,
            cost
        );

        const allowed = result[0] === 1;
        const remaining = result[1]; // Current tokens

        // Reset time is roughly when bucket is full? Or just 'window' seconds from now if empty?
        // For Token Bucket, 'reset' isn't as clear cut as "window resets at X".
        // We can estimate time to full: (capacity - current) / rate
        const timeToFull = (limit - remaining) / refillRate;
        const resetTime = Math.ceil(now + timeToFull);

        return {
            allowed,
            remaining: Math.floor(remaining),
            resetTime,
        };
    }
}

module.exports = TokenBucket;
